File: src/hover/latexComponent.ts
import { Language } from '../language';
import { getComponentMetadata } from '../metadata/component';
import * as range from '../range';
import { HoverProvider } from './provider';

export const LatexComponentHoverProvider: HoverProvider = {
  execute: async context => {
    const { document, params } = context;
    if (document.tree.language !== Language.Latex) {
      return undefined;
    }

    const include = document.tree.includes
      .filter(x => x.isUnitImport)
      .find(x => range.contains(x.command.range, params.position));

    if (include === undefined) {
      return undefined;
    }

    const metadata = await getComponentMetadata(include.path);
    if (metadata === undefined) {
      return undefined;
    }

    return {
      contents: metadata.documentation,
    };
  },
};

File: src/hover/latex_component.rs
use futures_boxed::boxed;
use texlab_protocol::*;
use texlab_syntax::*;
use texlab_workspace::*;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexComponentHoverProvider;

impl FeatureProvider for LatexComponentHoverProvider {
    type Params = TextDocumentPositionParams;
    type Output = Option<Hover>;

    #[boxed]
    async fn execute<'a>(
        &'a self,
        request: &'a FeatureRequest<TextDocumentPositionParams>,
    ) -> Option<Hover> {
        if let SyntaxTree::Latex(tree) = &request.document().tree {
            for include in &tree.includes {
                if include.kind == LatexIncludeKind::Package
                    || include.kind == LatexIncludeKind::Class
                {
                    for path in include.paths() {
                        if path.range().contains(request.params.position) {
                            let documentation = COMPLETION_DATABASE.documentation(path.text())?;
                            return Some(Hover {
                                contents: HoverContents::Markup(documentation),
                                range: Some(path.range()),
                            });
                        }
                    }
                }
            }
        }
        None
    }
}
